app/script/conversation/ConversationService.js

```diff
@@ -709,42 +709,72 @@ z.conversation.ConversationService = class ConversationService {
-                  update_version: changes.version,
-                };
-              }
-
-              Raygun.send('Failed sequential database update', custom_data);
-              throw new z.storage.StorageError(z.storage.StorageError.TYPE.NON_SEQUENTIAL_UPDATE);
-            });
-          });
-        }
+  updateMessageInDb(messageEntity, changes = {}, conversationId) {
+    return Promise.resolve(messageEntity.primary_key).then(primaryKey => {
+      const hasChanges = !!Object.keys(changes).length;
+      if (!hasChanges) {
+        throw new z.conversation.ConversationError(z.conversation.ConversationError.TYPE.NO_CHANGES);
+      }
+
+      const hasVersionedChanges = !!changes.version;
+      if (hasVersionedChanges) {
+        throw new z.conversation.ConversationError(z.conversation.ConversationError.TYPE.WRONG_CHANGE);
+      }
+
+      return this.storageService.update(this.EVENT_STORE_NAME, primaryKey, changes);
+    });
+  }
+
+  /**
+   * Update a message in the database.
+   *
+   * @param {Message} messageEntity - Message event to update in the database
+   * @param {Object} [changes={}] - Changes to update message with
+   * @param {string} conversationId - ID of conversation
+   * @returns {Promise} Resolves when the message was updated in database
+   */
+  updateMessageInDbByVersion(messageEntity, changes = {}, conversationId) {
+    return Promise.resolve(messageEntity.primary_key).then(primaryKey => {
+      const hasChanges = !!Object.keys(changes).length;
+      if (!hasChanges) {
+        throw new z.conversation.ConversationError(z.conversation.ConversationError.TYPE.NO_CHANGES);
+      }
 
-        return this.storageService.update(this.EVENT_STORE_NAME, primary_key, changes);
+      const hasVersionedChanges = !!changes.version;
+      if (!hasVersionedChanges) {
+        throw new z.conversation.ConversationError(z.conversation.ConversationError.TYPE.WRONG_CHANGE);
       }
 
-      throw new z.conversation.ConversationError(z.conversation.ConversationError.TYPE.NO_CHANGES);
+      return this.storageService.db.transaction('rw', this.EVENT_STORE_NAME, () => {
+        return this.load_event_from_db(conversationId, messageEntity.id).then(record => {
+          let customData;
+
+          if (record) {
+            const databaseVersion = record.version || 1;
+
+            const isSequentialUpdate = changes.version === databaseVersion + 1;
+            if (isSequentialUpdate) {
+              return this.storageService.update(this.EVENT_STORE_NAME, primaryKey, changes);
+            }
+
+            customData = {
+              databaseVersion: databaseVersion,
+              updateVersion: changes.version,
+            };
+          }
+
+          const logMessage = 'Failed sequential database update';
+          this.logger.error(logMessage, customData);
+
+          Raygun.send(new Error(logMessage), customData);
+          throw new z.storage.StorageError(z.storage.StorageError.TYPE.NON_SEQUENTIAL_UPDATE);
```

Comment:
Maybe we should differentiate between `non sequential updates` and `updates to a non existent entity`. I think the logs would be more meaningful like this.
WDYT?